sys/aarch64/rootdev.c

```diff
@@ -138,28 +139,34 @@ static void rootdev_disable_irq(intr_event_t *ie) {
   }
 }
 
-static void rootdev_intr_setup(device_t *dev, unsigned num,
-                               intr_handler_t *handler) {
+static void rootdev_intr_setup(device_t *dev, resource_t *r,
+                               ih_filter_t *filter, ih_service_t *service,
+                               void *arg, const char *name) {
   rootdev_t *rd = dev->parent->state;
+  int irq = r->r_start;
+  assert(irq < NIRQ);
+
+  if (rd->intr_event[irq] == NULL)
+    rd->intr_event[irq] = intr_event_create(dev, irq, rootdev_disable_irq,
+                                            rootdev_enable_irq, "???");
 
-  assert(num < NIRQ);
-  intr_event_t *event = &rd->intr_event[num];
-  intr_event_add_handler(event, handler);
+  r->r_handler =
+    intr_event_add_handler(rd->intr_event[irq], filter, service, arg, name);
 }
 
-static void rootdev_intr_teardown(device_t *dev, intr_handler_t *handler) {
-  intr_event_remove_handler(handler);
+static void rootdev_intr_teardown(device_t *dev, resource_t *irq) {
+  intr_event_remove_handler(irq->r_handler);
```

Comment:
We are leaking an interrupt handler here.

Reply:
Done.